### PR TITLE
SendGrid Delivery Events

### PR DESCRIPTION
**Pull request by jverce**
This PR is in support of issue #557 

* Refactor common code to be used in multiple components
* Split common code for timer-based vs http-based event sources
* Add event source logic for Delivery Events

### CODE REVIEW

File: components/sendgrid/sendgrid.app.js [hunk 1]
```diff
@@ -68,5 +72,23 @@ module.exports = {
         retriableStatusCodes,
       );
     },
+    async getWebhookSettings() {
+      const url = this._webhookSettingsUrl();
+      const requestConfig = this._makeRequestConfig();
+      const retriableStatusCodes = new Set([408, 429]);
```
Comment (tjk) [inline components/sendgrid/sendgrid.app.js:78]: nit but if these are always the same maybe just default in `withRetries`?

File: components/sendgrid/sources/common/http-based.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+module.exports = {
+  ...base,
+  props: {
+    ...base.props,
+    http: "$.interface.http",
+  },
+  hooks: {
+    ...base.hooks,
+    async activate() {
```
Comment (tjk) [inline components/sendgrid/sources/common/http-based.js:11]: I think deactive/activate get called anytime props are updated for example... can we be a bit more forgiving here if we are the owners of the webhook?

File: components/sendgrid/sources/common/http-based.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      throw new Error('webhookEventFlags is not implemented');
+    },
+    generateMeta(data) {
+      const {
+        sg_event_id: id,
+        timestamp: ts,
+      } = data;
+      return {
+        id,
+        summary: "New event",
```
Comment (tjk) [inline components/sendgrid/sources/common/http-based.js:63]: can we do a better summary?

Comment (tjk) [reply components/sendgrid/sources/common/http-based.js:63]: nevermind this is http-base

File: components/sendgrid/sources/delivery-events/delivery-events.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  name: "Delivery Events (Instant)",
+  description: "Emit an event when a delivery event is detected in SendGrid",
+  version: "0.0.1",
+  dedupe: "unique",
+  props: {
+    ...common.props,
+    processed: {
+      type: "boolean",
+      label: "Processed",
+      description: `
```
Comment (tjk) [inline components/sendgrid/sources/delivery-events/delivery-events.js:15]: hm i suppose these descriptions with newlines are probably okay for when embedded in html but maybe we shouldn't do them like this? (i suppose we could strip out in other clients maybe.. hm..)

**Review by tjk — APPROVED** [review]
mostly minor comments, the activate one is my main question

File: components/sendgrid/sources/delivery-events/delivery-events.js (hunk 4 above)
Comment (jverce) [reply components/sendgrid/sources/delivery-events/delivery-events.js:15]: That is actually I question I raised in another PR. If descriptions are a free-form text, and can be arbitrarily long, then at some point someone will add a very long line to the code.
🤷🏻‍♂️

File: components/sendgrid/sources/common/http-based.js (hunk 2 above)
Comment (jverce) [reply components/sendgrid/sources/common/http-based.js:11]: That's a good idea. We can check if the URL's match for example. Does that make sense?

File: components/sendgrid/sendgrid.app.js (hunk 1 above)
Comment (jverce) [reply components/sendgrid/sendgrid.app.js:78]: That's a good idea, looks like all API endpoints share these status codes, so we can reuse them.
Will fix!

**Pull request merged**